boa/src/builtins/string/mod.rs

```diff
@@ -238,17 +282,21 @@ impl String {
         // First we get it the actual string a private field stored on the object only the context has access to.
         // Then we convert it into a Rust String by wrapping it in from_value
         let primitive_val = this.to_string(context)?;
-
-        // Calling .len() on a string would give the wrong result, as they are bytes not the number of unicode code points
-        // Note that this is an O(N) operation (because UTF-8 is complex) while getting the number of bytes is an O(1) operation.
-        let length = primitive_val.chars().count();
         let pos = args
             .get(0)
             .cloned()
             .unwrap_or_else(Value::undefined)
             .to_integer(context)? as i32;
 
-        if pos >= length as i32 || pos < 0 {
+        // Fast path returning NaN when pos is obviously out of range
+        if pos < 0 || pos >= primitive_val.len() as i32 {
+            return Ok(Value::from(NAN));
+        }
+
+        // Calling .len() on a string would give the wrong result, as they are bytes not the number of unicode code points
+        // Note that this is an O(N) operation (because UTF-8 is complex) while getting the number of bytes is an O(1) operation.
+        let length = primitive_val.chars().count() as i32;
+        if pos >= length {
             return Ok(Value::from(NAN));
         }
```

Comment:
Like in `char_at`, this doesn't actually seem like a good idea, we could just match on `primitive_val.encode_utf16().nth(pos as usize)`, right?